Cleaning The Dataset


Data cleaning is important because, if data is wrong or contains missing values, 
outcomes and algorithms will be unreliable. 
So, we must clean it first before doing any process on the data.

In [1]:
# importing pandas to build a data frame from the csv file
import pandas as pd

In [2]:
# Importing the dataset as csv file
myData = pd.read_csv('companies.csv')

In [3]:
# Display the first ten rows of dataset 
myData.head(10)

ID          Name            Industry  Inception  Employees State  \
0   1      Over-Hex            Software     2006.0       25.0    TN   
1   2     Unimattax         IT Services     2009.0       36.0    PA   
2   3      Greenfax              Retail     2012.0        NaN    SC   
3   4     Blacklane         IT Services     2011.0       10.5    CA   
4   5      Yearflex            Software    20130.0       66.0    WI   
5   6  Indigoplanet         IT Services     2013.0       45.0    NJ   
6   7       Treslam  Financial Services     2009.0       60.0    MO   
7   8     Rednimdox        Construction     2013.0      116.0    NY   
8   9       Lamtone         IT Services    20090.0       73.0    CA   
9  10     Stripfind  Financial Services     2010.0       55.0    FL   

             City       Revenue           Expenses      Profit Growth  
0        Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1  Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
2      Greenville   $9,746,272   1,044,375 Dollars   8701897.0    16%  
3          Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4         Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5       Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6         Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
7        Woodside           NaN                NaN         NaN    NaN  
8       San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9      Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%

1.Detecting Null Values


In [4]:
# Display null values
myData.isna()

ID   Name  Industry  Inception  Employees  State   City  Revenue  \
0    False  False     False      False      False  False  False    False   
1    False  False     False      False      False  False  False    False   
2    False  False     False      False       True  False  False    False   
3    False  False     False      False      False  False  False    False   
4    False  False     False      False      False  False  False    False   
..     ...    ...       ...        ...        ...    ...    ...      ...   
496  False  False     False      False      False  False  False    False   
497  False  False     False      False      False  False  False    False   
498  False  False     False      False      False  False  False    False   
499  False  False     False      False      False  False  False    False   
500  False  False     False      False      False  False  False    False   

     Expenses  Profit  Growth  
0       False   False   False  
1       False   False   False  
2       False   False   False  
3       False   False   False  
4       False   False   False  
..        ...     ...     ...  
496     False   False   False  
497     False   False   False  
498     False   False   False  
499     False   False   False  
500     False   False   False  

[501 rows x 11 columns]

In [5]:
# Display the total number of the null values in every column
myData.isna().sum()

ID           0
Name         0
Industry     2
Inception    1
Employees    2
State        4
City         0
Revenue      2
Expenses     3
Profit       2
Growth       1
dtype: int64

In [6]:
myData.isna().sum().sum()

17

2.1. Filling Null Values

In [7]:
# Filling all null values with a static number such as 0
demo1 = myData.fillna(value=0)
demo1.head(10)

ID          Name            Industry  Inception  Employees State  \
0   1      Over-Hex            Software     2006.0       25.0    TN   
1   2     Unimattax         IT Services     2009.0       36.0    PA   
2   3      Greenfax              Retail     2012.0        0.0    SC   
3   4     Blacklane         IT Services     2011.0       10.5    CA   
4   5      Yearflex            Software    20130.0       66.0    WI   
5   6  Indigoplanet         IT Services     2013.0       45.0    NJ   
6   7       Treslam  Financial Services     2009.0       60.0    MO   
7   8     Rednimdox        Construction     2013.0      116.0    NY   
8   9       Lamtone         IT Services    20090.0       73.0    CA   
9  10     Stripfind  Financial Services     2010.0       55.0    FL   

             City       Revenue           Expenses      Profit Growth  
0        Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1  Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
2      Greenville   $9,746,272   1,044,375 Dollars   8701897.0    16%  
3          Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4         Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5       Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6         Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
7        Woodside             0                  0         0.0      0  
8       San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9      Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%

In [8]:
# Filling each null value with its previous value 
demo2 = myData.fillna(method='pad')
demo2.head(10)

ID          Name            Industry  Inception  Employees State  \
0   1      Over-Hex            Software     2006.0       25.0    TN   
1   2     Unimattax         IT Services     2009.0       36.0    PA   
2   3      Greenfax              Retail     2012.0       36.0    SC   
3   4     Blacklane         IT Services     2011.0       10.5    CA   
4   5      Yearflex            Software    20130.0       66.0    WI   
5   6  Indigoplanet         IT Services     2013.0       45.0    NJ   
6   7       Treslam  Financial Services     2009.0       60.0    MO   
7   8     Rednimdox        Construction     2013.0      116.0    NY   
8   9       Lamtone         IT Services    20090.0       73.0    CA   
9  10     Stripfind  Financial Services     2010.0       55.0    FL   

             City       Revenue           Expenses      Profit Growth  
0        Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1  Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
2      Greenville   $9,746,272   1,044,375 Dollars   8701897.0    16%  
3          Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4         Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5       Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6         Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
7        Woodside   $5,387,469   2,127,984 Dollars   3259485.0    17%  
8       San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9      Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%

In [9]:
# Filling each null value with its following value 
demo3 = myData.fillna(method= 'bfill')
demo3.head(10)

ID          Name            Industry  Inception  Employees State  \
0   1      Over-Hex            Software     2006.0       25.0    TN   
1   2     Unimattax         IT Services     2009.0       36.0    PA   
2   3      Greenfax              Retail     2012.0       10.5    SC   
3   4     Blacklane         IT Services     2011.0       10.5    CA   
4   5      Yearflex            Software    20130.0       66.0    WI   
5   6  Indigoplanet         IT Services     2013.0       45.0    NJ   
6   7       Treslam  Financial Services     2009.0       60.0    MO   
7   8     Rednimdox        Construction     2013.0      116.0    NY   
8   9       Lamtone         IT Services    20090.0       73.0    CA   
9  10     Stripfind  Financial Services     2010.0       55.0    FL   

             City       Revenue           Expenses      Profit Growth  
0        Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1  Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
2      Greenville   $9,746,272   1,044,375 Dollars   8701897.0    16%  
3          Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4         Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5       Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6         Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
7        Woodside  $11,757,018   6,482,465 Dollars   5274553.0    30%  
8       San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9      Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%

In [10]:
# Filling numeric values with the mean value in a specific column
mn = myData["Employees"].mean()
myData['Employees'].fillna(mn, inplace=True)
myData.head(10)

ID          Name            Industry  Inception   Employees State  \
0   1      Over-Hex            Software     2006.0   25.000000    TN   
1   2     Unimattax         IT Services     2009.0   36.000000    PA   
2   3      Greenfax              Retail     2012.0  148.566132    SC   
3   4     Blacklane         IT Services     2011.0   10.500000    CA   
4   5      Yearflex            Software    20130.0   66.000000    WI   
5   6  Indigoplanet         IT Services     2013.0   45.000000    NJ   
6   7       Treslam  Financial Services     2009.0   60.000000    MO   
7   8     Rednimdox        Construction     2013.0  116.000000    NY   
8   9       Lamtone         IT Services    20090.0   73.000000    CA   
9  10     Stripfind  Financial Services     2010.0   55.000000    FL   

             City       Revenue           Expenses      Profit Growth  
0        Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1  Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
2      Greenville   $9,746,272   1,044,375 Dollars   8701897.0    16%  
3          Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4         Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5       Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6         Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
7        Woodside           NaN                NaN         NaN    NaN  
8       San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9      Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%

In [11]:
myData = pd.read_csv('companies.csv')

# Filling numeric values with the median value in a specific column
med = myData["Employees"].median()
myData["Employees"].fillna(med, inplace = True)
myData.head(10)

ID          Name            Industry  Inception  Employees State  \
0   1      Over-Hex            Software     2006.0       25.0    TN   
1   2     Unimattax         IT Services     2009.0       36.0    PA   
2   3      Greenfax              Retail     2012.0       56.0    SC   
3   4     Blacklane         IT Services     2011.0       10.5    CA   
4   5      Yearflex            Software    20130.0       66.0    WI   
5   6  Indigoplanet         IT Services     2013.0       45.0    NJ   
6   7       Treslam  Financial Services     2009.0       60.0    MO   
7   8     Rednimdox        Construction     2013.0      116.0    NY   
8   9       Lamtone         IT Services    20090.0       73.0    CA   
9  10     Stripfind  Financial Services     2010.0       55.0    FL   

             City       Revenue           Expenses      Profit Growth  
0        Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1  Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
2      Greenville   $9,746,272   1,044,375 Dollars   8701897.0    16%  
3          Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4         Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5       Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6         Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
7        Woodside           NaN                NaN         NaN    NaN  
8       San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9      Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%

In [12]:
myData = pd.read_csv('companies.csv')

# Filling numeric values with the mode value in a specific column
mod = myData["Employees"].mode()[0]
myData["Employees"].fillna(mod, inplace = True)
myData.head(10)

ID          Name            Industry  Inception  Employees State  \
0   1      Over-Hex            Software     2006.0       25.0    TN   
1   2     Unimattax         IT Services     2009.0       36.0    PA   
2   3      Greenfax              Retail     2012.0       25.0    SC   
3   4     Blacklane         IT Services     2011.0       10.5    CA   
4   5      Yearflex            Software    20130.0       66.0    WI   
5   6  Indigoplanet         IT Services     2013.0       45.0    NJ   
6   7       Treslam  Financial Services     2009.0       60.0    MO   
7   8     Rednimdox        Construction     2013.0      116.0    NY   
8   9       Lamtone         IT Services    20090.0       73.0    CA   
9  10     Stripfind  Financial Services     2010.0       55.0    FL   

             City       Revenue           Expenses      Profit Growth  
0        Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1  Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
2      Greenville   $9,746,272   1,044,375 Dollars   8701897.0    16%  
3          Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4         Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5       Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6         Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
7        Woodside           NaN                NaN         NaN    NaN  
8       San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9      Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%

2.2. Drop All Null Values Instead Of Filling It

In [13]:
myData = pd.read_csv('companies.csv')

# Drop all null values
data = myData.dropna()
data.head(10)

ID          Name            Industry  Inception  Employees State  \
0    1      Over-Hex            Software     2006.0       25.0    TN   
1    2     Unimattax         IT Services     2009.0       36.0    PA   
3    4     Blacklane         IT Services     2011.0       10.5    CA   
4    5      Yearflex            Software    20130.0       66.0    WI   
5    6  Indigoplanet         IT Services     2013.0       45.0    NJ   
6    7       Treslam  Financial Services     2009.0       60.0    MO   
8    9       Lamtone         IT Services    20090.0       73.0    CA   
9   10     Stripfind  Financial Services     2010.0       55.0    FL   
11  12      Mattouch         IT Services     2013.0        6.0    WA   
12  13     Techdrill              Health     2009.0        6.0    MS   

              City       Revenue           Expenses      Profit Growth  
0         Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1   Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
3           Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4          Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5        Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6          Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
8        San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9       Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%  
11        Bellevue  $14,026,934   7,429,377 Dollars   6597557.0    26%  
12         Flowood  $10,573,990   7,435,363 Dollars   3138627.0     8%

In [14]:
# Display the null values detection in each column
data.isna().sum()


ID           0
Name         0
Industry     0
Inception    0
Employees    0
State        0
City         0
Revenue      0
Expenses     0
Profit       0
Growth       0
dtype: int64

3. Wrong Formatted Data


Note: null values can be considered as wrong formatted data That we can easily fill or drop from the dataset as previous.

We can solve the wrong formatted data by converting the column into the same data format. 
In our dataset, we have a float number in the 'Employees' column and its third row (after dropping nulls) has 10.5 employees which is not logical. So, we convert it to integer format to solve this logical error.  

In [15]:
data['Employees'] = data['Employees'].astype(int)

# The 'Inception' year also must be integer, so we convert it also.
data['Inception'] = data['Inception'].astype(int)
data.head(10)

C:\Users\AHMADK~1\AppData\Local\Temp/ipykernel_12736/1159873316.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Employees'] = data['Employees'].astype(int)
C:\Users\AHMADK~1\AppData\Local\Temp/ipykernel_12736/1159873316.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Inception'] = data['Inception'].astype(int)


ID          Name            Industry  Inception  Employees State  \
0    1      Over-Hex            Software       2006         25    TN   
1    2     Unimattax         IT Services       2009         36    PA   
3    4     Blacklane         IT Services       2011         10    CA   
4    5      Yearflex            Software      20130         66    WI   
5    6  Indigoplanet         IT Services       2013         45    NJ   
6    7       Treslam  Financial Services       2009         60    MO   
8    9       Lamtone         IT Services      20090         73    CA   
9   10     Stripfind  Financial Services       2010         55    FL   
11  12      Mattouch         IT Services       2013          6    WA   
12  13     Techdrill              Health       2009          6    MS   

              City       Revenue           Expenses      Profit Growth  
0         Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1   Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
3           Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4          Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5        Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6          Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
8        San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9       Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%  
11        Bellevue  $14,026,934   7,429,377 Dollars   6597557.0    26%  
12         Flowood  $10,573,990   7,435,363 Dollars   3138627.0     8%

4. Wrong Data

It needn't be null values or wrong formatted data, it can be wrong like if someone writes "20000" instead of "2000" in the Inception year, although we are still in the 2022 year.
if the data set is small, you can look at it and change the wrong data one by one, but the huge ones need some sophisticated code blocks to detect and correct it, or you can drop it.

In our dataset, we note that the fifth and the ninth rows have wrong data in the 'Inception' column so, we need to correct it. 

In [16]:
data.loc[4, 'Inception'] = 2013
data.loc[8, 'Inception'] = 2009
data.head(10)

c:\Users\AhmadKhabeer\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\core\indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


ID          Name            Industry  Inception  Employees State  \
0    1      Over-Hex            Software       2006         25    TN   
1    2     Unimattax         IT Services       2009         36    PA   
3    4     Blacklane         IT Services       2011         10    CA   
4    5      Yearflex            Software       2013         66    WI   
5    6  Indigoplanet         IT Services       2013         45    NJ   
6    7       Treslam  Financial Services       2009         60    MO   
8    9       Lamtone         IT Services       2009         73    CA   
9   10     Stripfind  Financial Services       2010         55    FL   
11  12      Mattouch         IT Services       2013          6    WA   
12  13     Techdrill              Health       2009          6    MS   

              City       Revenue           Expenses      Profit Growth  
0         Franklin   $9,684,527   1,130,700 Dollars   8553827.0    19%  
1   Newtown Square  $14,016,543     804,035 Dollars  13212508.0    20%  
3           Orange  $15,359,369   4,631,808 Dollars  10727561.0    19%  
4          Madison   $8,567,910   4,374,841 Dollars   4193069.0    19%  
5        Manalapan  $12,805,452   4,626,275 Dollars   8179177.0    22%  
6          Clayton   $5,387,469   2,127,984 Dollars   3259485.0    17%  
8        San Ramon  $11,757,018   6,482,465 Dollars   5274553.0    30%  
9       Boca Raton  $12,329,371     916,455 Dollars  11412916.0    20%  
11        Bellevue  $14,026,934   7,429,377 Dollars   6597557.0    26%  
12         Flowood  $10,573,990   7,435,363 Dollars   3138627.0     8%

5. Removing Duplicates

Duplicated rows are more than one row that has the same data. That can affect the processing of the data badly.

We need to check if there are any duplicates or not using duplicated() that return 'True' if the row is duplicated and 'False' otherwise. 

In [19]:
data.duplicated()

0      False
1      False
3      False
4      False
5      False
       ...  
496    False
497    False
498    False
499    False
500     True
Length: 489, dtype: bool

We note that the row with ID 500 is duplicated. So, we need to drop the duplicated row

In [20]:
data.drop_duplicates(inplace = True)
data.duplicated()

c:\Users\AhmadKhabeer\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\util\_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


0      False
1      False
3      False
4      False
5      False
       ...  
495    False
496    False
497    False
498    False
499    False
Length: 488, dtype: bool

So, finally we can say that we clean the dataset and now we can move to the next step.